staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go

```diff
@@ -0,0 +1,257 @@
+	rap.ascendingIndices = make([]int, len(rap.items)*2)
+	for idx := 0; idx < len(rap.ascendingIndices); idx++ {
+		rap.ascendingIndices[idx] = idx
+	}
+	sort.Sort(rap)
+	return rap
+}
+
+func (rap *relativeAllocProblem) getItemIndex(idx int) (int, bool) {
+	packedIndex := rap.ascendingIndices[idx]
+	itemIndex := packedIndex / 2
+	return itemIndex, packedIndex == itemIndex*2
+}
+
+// decode(J) returns the bound associated with ascendingIndices[J], the associated items index,
+// and a bool indicating whether the bound is the item's lower bound.
+func (rap *relativeAllocProblem) decode(idx int) (float32, int, bool) {
+	itemIdx, lower := rap.getItemIndex(idx)
+	if lower {
+		return rap.items[itemIdx].relativeLowerBound, itemIdx, lower
+	}
+	return rap.items[itemIdx].relativeUpperBound, itemIdx, lower
+}
+
+func (rap *relativeAllocProblem) getProportion(idx int) float32 {
+	prop, _, _ := rap.decode(idx)
+	return prop
+}
+
+func (rap *relativeAllocProblem) Len() int { return len(rap.items) * 2 }
+
+func (rap *relativeAllocProblem) Less(i, j int) bool {
+	return rap.getProportion(i) < rap.getProportion(j)
+}
+
+func (rap *relativeAllocProblem) Swap(i, j int) {
+	rap.ascendingIndices[i], rap.ascendingIndices[j] = rap.ascendingIndices[j], rap.ascendingIndices[i]
+}
+
+// minMax records the minimum and maximum value seen while scanning a set of numbers
+type minMax struct {
+	min float32
+	max float32
+}
+
+// note scans one more number
+func (mm *minMax) note(x float32) {
+	mm.min = f32min(mm.min, x)
+	mm.max = f32max(mm.max, x)
+}
+
+// computeConcurrencyAllocation returns the unique `allocs []float32`, and
+// an associated `fairProp float32`, that jointly have
+// all of the following properties (to the degree that floating point calculations allow)
+// if possible otherwise returns an error saying why it is impossible.
+// `allocs` sums to `requiredSum`.
+// For each J in [0, len(classes)):
+// (1) `classes[J].lowerBound <= allocs[J] <= classes[J].upperBound` and
+// (2) exactly one of the following is true:
+// (2a) `allocs[J] == fairProp * classes[J].target`,
+// (2b) `allocs[J] == classes[J].lowerBound && classes[J].lowerBound > fairProp * classes[J].target`, or
+// (2c) `allocs[J] == classes[J].upperBound && classes[J].upperBound < fairProp * classes[J].target`.
+// Each allocProblemItem is required to have `target >= lowerBound` and `upperBound > 0`.
+// A target of zero is treated as if it were 0.001.
+func computeConcurrencyAllocation(requiredSum int, classes []allocProblemItem) ([]float32, float32, error) {
+	if requiredSum < 0 {
+		return nil, 0, errors.New("negative sums are not supported")
+	}
+	requiredSumF := float32(requiredSum)
+	var lowSum, highSum, targetSum float32
+	ubRange := minMax{min: float32(math.MaxFloat32)}
+	lbRange := minMax{min: float32(math.MaxFloat32)}
+	relativeItems := make([]relativeAllocItem, len(classes))
+	for idx, item := range classes {
+		target := item.target
+		if target < item.lowerBound {
```

Comment:
The target can be greater than the upper bound.  The KEP explains the reason for the asymmetric imposition of bounds on the target.  In https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1040-priority-and-fairness#dispatching look for the following.

> This shares the wealth or the pain proportionally among the priority levels (but note: the upper bound does not affect the target, lest the pain of not achieving a high SmoothSeatDemand be distorted, while the lower bound does affect the target, so that merely achieving the lower bound is not considered a gain).